lib/facter/resolvers/containers.rb

```diff
@@ -14,45 +14,59 @@ class << self
         private
 
         def post_resolve(fact_name, _options)
-          @fact_list.fetch(fact_name) { read_cgroup(fact_name) }
+          @fact_list.fetch(fact_name) do
+            read_environ(fact_name) || read_cgroup(fact_name)
+          end
         end
 
         def read_cgroup(fact_name)
           output_cgroup = Facter::Util::FileHelper.safe_read('/proc/1/cgroup', nil)
-          output_environ = Facter::Util::FileHelper.safe_read('/proc/1/environ', nil)
-          return unless output_cgroup && output_environ
+          return unless output_cgroup
 
           output_docker = %r{docker/(.+)}.match(output_cgroup)
           output_lxc = %r{^/lxc/([^/]+)}.match(output_cgroup)
-          lxc_from_environ = /container=lxc/ =~ output_environ
 
-          info, vm = extract_vm_and_info(output_docker, output_lxc, lxc_from_environ)
-          info, vm = extract_for_nspawn(output_environ) unless vm
+          info, vm = extract_vm_and_info(output_docker, output_lxc)
           @fact_list[:vm] = vm
           @fact_list[:hypervisor] = { vm.to_sym => info } if vm
           @fact_list[fact_name]
         end
 
-        def extract_vm_and_info(output_docker, output_lxc, lxc_from_environ)
+        def read_environ(fact_name)
+          container = Facter::Util::Linux::Proc.getenv_for_pid(1, 'container')
+          return unless container && container != ''
+
+          info = {}
+          case container
+          when 'lxcroot'
+            vm = 'lxc'
+          when 'podman'
+            vm = 'podman'
+          when 'crio'
+            vm = 'crio'
+          when 'systemd-nspawn'
+            vm = 'systemd_nspawn'
+            info = { 'id' => Facter::Util::FileHelper.safe_read('/etc/machine-id', nil).strip }
+          else
+            vm = 'container_other'
```

Comment:
Could you add a test for this case? Should we log a warning that it's an unknown container?

Reply:
yup, added.